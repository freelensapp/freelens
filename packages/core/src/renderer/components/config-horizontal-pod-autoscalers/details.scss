/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

@use "autoscaler.mixins";
@use "../vars" as *;

.HpaDetails {
  .status {
    @include autoscaler.hpa-status-bgc;
  }

  .metrics .Table {
    margin: 0 (-$margin * 3);

    .TableCell {
      word-break: break-word;


      &:first-child {
        margin-left: $margin * 2;
      }

      &:last-child {
        margin-right: $margin * 2;
      }

      &.name {
        flex-grow: 2;
      }
    }
  }
}
