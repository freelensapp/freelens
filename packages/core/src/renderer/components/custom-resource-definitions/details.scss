/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

@use "mixins.scss";
@use "../vars" as *;

.CustomResourceDefinitionDetails {
  .conditions {
    @include mixins.custom-resource-definitions-condition-colors;
  }

  .Table {
    margin-left: -$margin * 2;
    margin-right: -$margin * 2;

    .TableRow {
      cursor: default;
    }

    &.printer-columns {
      .type {
        flex: 0.5;
      }

      .json-path {
        flex: 3;
      }
    }
  }
}
