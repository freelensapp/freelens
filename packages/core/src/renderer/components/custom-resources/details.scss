/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

.CustomResourceDetails {
  .status {
    .ready {
      color: white;
      background-color: var(--colorOk);
    }
  }
}
