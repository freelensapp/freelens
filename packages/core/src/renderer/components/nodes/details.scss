/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

@use "nodes-mixins";

.NodeDetails {
  --status-ready-bg: var(--colorOk);
  --status-default-bg: var(--colorError);

  .conditions {
    @include nodes-mixins.node-status-bgs;
  }
}
