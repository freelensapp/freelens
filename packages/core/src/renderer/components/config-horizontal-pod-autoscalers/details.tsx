/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "./details.scss";

import type { HorizontalPodAutoscalerMetricSpec, HorizontalPodAutoscalerMetricTarget } from "@freelensapp/kube-object";
import { HorizontalPodAutoscaler } from "@freelensapp/kube-object";
import type { Logger } from "@freelensapp/logger";
import { loggerInjectionToken } from "@freelensapp/logger";
import { cssNames } from "@freelensapp/utilities";
import { withInjectables } from "@ogre-tools/injectable-react";
import { observer } from "mobx-react";
import React from "react";
import { Link } from "react-router-dom";
import type { ApiManager } from "../../../common/k8s-api/api-manager";
import apiManagerInjectable from "../../../common/k8s-api/api-manager/manager.injectable";
import { Badge } from "../badge";
import { DrawerItem, DrawerTitle } from "../drawer";
import type { GetDetailsUrl } from "../kube-detail-params/get-details-url.injectable";
import getDetailsUrlInjectable from "../kube-detail-params/get-details-url.injectable";
import type { KubeObjectDetailsProps } from "../kube-object-details";
import { Table, TableCell, TableHead, TableRow } from "../table";
import { getMetricName } from "./get-metric-name";
import getHorizontalPodAutoscalerMetrics from "./get-metrics.injectable";

export interface HpaDetailsProps extends KubeObjectDetailsProps<HorizontalPodAutoscaler> {}

interface Dependencies {
  apiManager: ApiManager;
  logger: Logger;
  getDetailsUrl: GetDetailsUrl;
  getMetrics: (hpa: HorizontalPodAutoscaler) => string[];
}

@observer
class NonInjectedHorizontalPodAutoscalerDetails extends React.Component<HpaDetailsProps & Dependencies> {
  private renderTargetLink(target: HorizontalPodAutoscalerMetricTarget | undefined) {
    if (!target) {
      return null;
    }

    const { object: hpa, apiManager, getDetailsUrl } = this.props;
    const { kind, name } = target;
    const objectUrl = getDetailsUrl(apiManager.lookupApiLink(target, hpa));

    return (
      <>
        on
        <Link to={objectUrl}>{`${kind}/${name}`}</Link>
      </>
    );
  }

  renderMetrics() {
    const { object: hpa } = this.props;

    const renderName = (metric: HorizontalPodAutoscalerMetricSpec) => {
      const metricName = getMetricName(metric);

      switch (metric?.type) {
        case "ContainerResource":

        // fallthrough
        case "Resource": {
          const metricSpec = metric.resource ?? metric.containerResource;

          return `Resource ${metricSpec.name} on Pods`;
        }
        case "Pods":
          return `${metricName ?? ""} on Pods`;

        case "Object": {
          return (
            <>
              {metricName} {this.renderTargetLink(metric.object?.describedObject)}
            </>
          );
        }
        case "External":
          return `${metricName ?? ""} on ${JSON.stringify(metric.external.metricSelector ?? metric.external.metric?.selector)}`;
        default:
          return hpa.spec?.targetCPUUtilizationPercentage ? "CPU Utilization percentage" : "unknown";
      }
    };

    return (
      <Table data-testid="hpa-metrics">
        <TableHead flat>
          <TableCell className="name">Name</TableCell>
          <TableCell className="metrics">Current / Target</TableCell>
        </TableHead>
        {this.props.getMetrics(hpa).map((metrics, index) => (
          <TableRow key={index}>
            <TableCell className="name">{renderName(hpa.getMetrics()[index])}</TableCell>
            <TableCell className="metrics">{metrics}</TableCell>
          </TableRow>
        ))}
      </Table>
    );
  }

  render() {
    const { object: hpa, apiManager, getDetailsUrl, logger } = this.props;

    if (!hpa) {
      return null;
    }

    if (!(hpa instanceof HorizontalPodAutoscaler)) {
      logger.error("[HpaDetails]: passed object that is not an instanceof HorizontalPodAutoscaler", hpa);

      return null;
    }

    const { scaleTargetRef } = hpa.spec;

    return (
      <div className="HpaDetails">
        <DrawerItem name="Reference">
          {scaleTargetRef && (
            <Link to={getDetailsUrl(apiManager.lookupApiLink(scaleTargetRef, hpa))}>
              {scaleTargetRef.kind}/{scaleTargetRef.name}
            </Link>
          )}
        </DrawerItem>

        <DrawerItem name="Min Pods">{hpa.getMinPods()}</DrawerItem>

        <DrawerItem name="Max Pods">{hpa.getMaxPods()}</DrawerItem>

        <DrawerItem name="Replicas">{hpa.getReplicas()}</DrawerItem>

        <DrawerItem name="Status" className="status" labelsOnly>
          {hpa.getReadyConditions().map(({ type, tooltip, isReady }) => (
            <Badge key={type} label={type} tooltip={tooltip} className={cssNames({ [type.toLowerCase()]: isReady })} />
          ))}
        </DrawerItem>

        {(hpa.getMetrics().length !== 0 || hpa.spec?.targetCPUUtilizationPercentage) && (
          <>
            <DrawerTitle>Metrics</DrawerTitle>
            <div className="metrics">{this.renderMetrics()}</div>
          </>
        )}
      </div>
    );
  }
}

export const HorizontalPodAutoscalerDetails = withInjectables<Dependencies, HpaDetailsProps>(
  NonInjectedHorizontalPodAutoscalerDetails,
  {
    getProps: (di, props) => ({
      ...props,
      apiManager: di.inject(apiManagerInjectable),
      getDetailsUrl: di.inject(getDetailsUrlInjectable),
      logger: di.inject(loggerInjectionToken),
      getMetrics: di.inject(getHorizontalPodAutoscalerMetrics),
    }),
  },
);
