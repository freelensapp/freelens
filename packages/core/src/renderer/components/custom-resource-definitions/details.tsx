/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "./details.scss";

import { CustomResourceDefinition } from "@freelensapp/kube-object";
import type { Logger } from "@freelensapp/logger";
import { loggerInjectionToken } from "@freelensapp/logger";
import { withInjectables } from "@ogre-tools/injectable-react";
import { observer } from "mobx-react";
import React from "react";
import { Link } from "react-router-dom";
import { Badge } from "../badge";
import { DrawerItem, DrawerTitle } from "../drawer";
import { Input } from "../input";
import type { KubeObjectDetailsProps } from "../kube-object-details";
import { MonacoEditor } from "../monaco-editor";
import { Table, TableCell, TableHead, TableRow } from "../table";

export interface CustomResourceDefinitionDetailsProps extends KubeObjectDetailsProps<CustomResourceDefinition> {}

interface Dependencies {
  logger: Logger;
}

@observer
class NonInjectedCustomResourceDefinitionDetails extends React.Component<
  CustomResourceDefinitionDetailsProps & Dependencies
> {
  render() {
    const { object: crd } = this.props;

    if (!crd) {
      return null;
    }

    if (!(crd instanceof CustomResourceDefinition)) {
      this.props.logger.error(
        "[CustomResourceDefinitionDetails]: passed object that is not an instanceof CustomResourceDefinition",
        crd,
      );

      return null;
    }

    const { plural, singular, kind, listKind } = crd.getNames();
    const printerColumns = crd.getPrinterColumns();
    const validation = crd.getValidation();

    return (
      <div className="CustomResourceDefinitionDetails">
        <DrawerItem name="Group">{crd.getGroup()}</DrawerItem>
        <DrawerItem name="Version">{crd.getVersion()}</DrawerItem>
        <DrawerItem name="Stored versions">{crd.getStoredVersions()}</DrawerItem>
        <DrawerItem name="Scope">{crd.getScope()}</DrawerItem>
        <DrawerItem name="Resource">
          <Link to={crd.getResourceUrl()}>{crd.getResourceTitle()}</Link>
        </DrawerItem>
        <DrawerItem name="Conversion" className="flex gaps align-flex-start">
          <Input multiLine theme="round-black" className="box grow" value={crd.getConversion()} readOnly />
        </DrawerItem>
        <DrawerItem name="Conditions" className="conditions" labelsOnly>
          {crd.getConditions().map((condition) => {
            const { type, message, lastTransitionTime, status } = condition;

            return (
              <Badge
                key={type}
                label={type}
                disabled={status === "False"}
                className={type}
                tooltip={
                  <>
                    <p>{message}</p>
                    <p>
                      Last transition time:
                      {lastTransitionTime}
                    </p>
                  </>
                }
              />
            );
          })}
        </DrawerItem>
        <DrawerTitle>Names</DrawerTitle>
        <Table selectable className="names box grow">
          <TableHead>
            <TableCell>plural</TableCell>
            <TableCell>singular</TableCell>
            <TableCell>kind</TableCell>
            <TableCell>listKind</TableCell>
          </TableHead>
          <TableRow>
            <TableCell>{plural}</TableCell>
            <TableCell>{singular}</TableCell>
            <TableCell>{kind}</TableCell>
            <TableCell>{listKind}</TableCell>
          </TableRow>
        </Table>
        {printerColumns.length > 0 && (
          <>
            <DrawerTitle>Additional Printer Columns</DrawerTitle>
            <Table selectable className="printer-columns box grow">
              <TableHead>
                <TableCell className="name">Name</TableCell>
                <TableCell className="type">Type</TableCell>
                <TableCell className="json-path">JSON Path</TableCell>
              </TableHead>
              {printerColumns.map((column, index) => {
                const { name, type, jsonPath } = column;

                return (
                  <TableRow key={index}>
                    <TableCell className="name">{name}</TableCell>
                    <TableCell className="type">{type}</TableCell>
                    <TableCell className="json-path">
                      <Badge label={jsonPath} />
                    </TableCell>
                  </TableRow>
                );
              })}
            </Table>
          </>
        )}
        {validation && (
          <>
            <DrawerTitle>Validation</DrawerTitle>
            <MonacoEditor readOnly value={validation} style={{ height: 400 }} />
          </>
        )}
      </div>
    );
  }
}

export const CustomResourceDefinitionDetails = withInjectables<Dependencies, CustomResourceDefinitionDetailsProps>(
  NonInjectedCustomResourceDefinitionDetails,
  {
    getProps: (di, props) => ({
      ...props,
      logger: di.inject(loggerInjectionToken),
    }),
  },
);
