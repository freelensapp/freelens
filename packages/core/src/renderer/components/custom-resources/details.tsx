/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "./details.scss";

import type { AdditionalPrinterColumnsV1, KubeObjectMetadata, KubeObjectStatus } from "@freelensapp/kube-object";
import { CustomResourceDefinition, KubeObject } from "@freelensapp/kube-object";
import type { Logger } from "@freelensapp/logger";
import { loggerInjectionToken } from "@freelensapp/logger";
import type { StrictReactNode } from "@freelensapp/utilities";
import { cssNames, safeJSONPathValue } from "@freelensapp/utilities";
import { withInjectables } from "@ogre-tools/injectable-react";
import { observer } from "mobx-react";
import React from "react";
import { Badge } from "../badge";
import { DrawerItem } from "../drawer";
import { Input } from "../input";
import type { KubeObjectDetailsProps } from "../kube-object-details";

export interface CustomResourceDetailsProps extends KubeObjectDetailsProps<KubeObject> {
  crd?: CustomResourceDefinition;
}

function convertSpecValue(value: unknown): StrictReactNode {
  if (Array.isArray(value)) {
    return (
      <ul>
        {value.map((value, index) => (
          <li key={index}>{convertSpecValue(value)}</li>
        ))}
      </ul>
    );
  }

  if (typeof value === "object") {
    return <Input readOnly multiLine theme="round-black" className="box grow" value={JSON.stringify(value, null, 2)} />;
  }

  if (typeof value === "boolean" || typeof value === "string" || typeof value === "number") {
    return value.toString();
  }

  return null;
}

interface Dependencies {
  logger: Logger;
}

@observer
class NonInjectedCustomResourceDetails extends React.Component<CustomResourceDetailsProps & Dependencies> {
  renderAdditionalColumns(resource: KubeObject, columns: AdditionalPrinterColumnsV1[]) {
    return columns.map(({ name, jsonPath }) => (
      <DrawerItem key={name} name={name}>
        {convertSpecValue(safeJSONPathValue(resource, jsonPath))}
      </DrawerItem>
    ));
  }

  renderStatus(cr: KubeObject, columns: AdditionalPrinterColumnsV1[]) {
    const customResource = cr as KubeObject<KubeObjectMetadata, KubeObjectStatus, unknown>;
    const showStatus =
      !columns.find((column) => column.name == "Status") && Array.isArray(customResource.status?.conditions);

    if (!showStatus) {
      return null;
    }

    const conditions = customResource.status?.conditions
      ?.filter(({ type, reason }) => type || reason)
      .map(({ type, reason, message, status }) => ({
        kind: type || reason || "<unknown>",
        message,
        status,
      }))
      .map(({ kind, message, status }, index) => (
        <Badge
          key={`${kind}${index}`}
          label={kind}
          disabled={status === "False"}
          className={kind.toLowerCase()}
          tooltip={message}
        />
      ));

    return (
      <DrawerItem name="Status" className="status" labelsOnly>
        {conditions}
      </DrawerItem>
    );
  }

  render() {
    const {
      props: { object, crd, logger },
    } = this;

    if (!object || !crd) {
      return null;
    }

    if (!(object instanceof KubeObject)) {
      logger.error("[CrdResourceDetails]: passed object that is not an instanceof KubeObject", object);

      return null;
    }

    if (!(crd instanceof CustomResourceDefinition)) {
      logger.error("[CrdResourceDetails]: passed crd that is not an instanceof CustomResourceDefinition", crd);

      return null;
    }

    const extraColumns = crd.getPrinterColumns();

    return (
      <div className={cssNames("CustomResourceDetails", crd.getResourceKind())}>
        {this.renderAdditionalColumns(object, extraColumns)}
        {this.renderStatus(object, extraColumns)}
      </div>
    );
  }
}

export const CustomResourceDetails = withInjectables<Dependencies, CustomResourceDetailsProps>(
  NonInjectedCustomResourceDetails,
  {
    getProps: (di, props) => ({
      ...props,
      logger: di.inject(loggerInjectionToken),
    }),
  },
);
