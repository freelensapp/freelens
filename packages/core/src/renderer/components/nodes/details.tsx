/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "./details.scss";

import { Node, formatNodeTaint } from "@freelensapp/kube-object";
import type { Logger } from "@freelensapp/logger";
import { loggerInjectionToken } from "@freelensapp/logger";
import { withInjectables } from "@ogre-tools/injectable-react";
import kebabCase from "lodash/kebabCase";
import upperFirst from "lodash/upperFirst";
import { disposeOnUnmount, observer } from "mobx-react";
import React from "react";
import type { SubscribeStores } from "../../kube-watch-api/kube-watch-api";
import subscribeStoresInjectable from "../../kube-watch-api/subscribe-stores.injectable";
import { Badge } from "../badge";
import { DrawerItem } from "../drawer";
import { DrawerTitle } from "../drawer/drawer-title";
import type { KubeObjectDetailsProps } from "../kube-object-details";
import loadPodsFromAllNamespacesInjectable from "../workloads-pods/load-pods-from-all-namespaces.injectable";
import { PodDetailsList } from "../workloads-pods/pod-details-list";
import type { PodStore } from "../workloads-pods/store";
import podStoreInjectable from "../workloads-pods/store.injectable";
import { NodeDetailsResources } from "./details-resources";

export interface NodeDetailsProps extends KubeObjectDetailsProps<Node> {}

interface Dependencies {
  subscribeStores: SubscribeStores;
  podStore: PodStore;
  logger: Logger;
  loadPodsFromAllNamespaces: () => void;
}

@observer
class NonInjectedNodeDetails extends React.Component<NodeDetailsProps & Dependencies> {
  componentDidMount() {
    disposeOnUnmount(this, [this.props.subscribeStores([this.props.podStore])]);

    this.props.loadPodsFromAllNamespaces();
  }

  render() {
    const { object: node, podStore, logger } = this.props;

    if (!node) {
      return null;
    }

    if (!(node instanceof Node)) {
      logger.error("[NodeDetails]: passed object that is not an instanceof Node", node);

      return null;
    }

    const { nodeInfo, addresses } = node.status ?? {};
    const conditions = node.getActiveConditions();
    const taints = node.getTaints();
    const childPods = podStore.getPodsByNode(node.getName());

    return (
      <div className="NodeDetails">
        {addresses && (
          <DrawerItem name="Addresses">
            {addresses.map(({ type, address }) => (
              <p key={type}>{`${type}: ${address}`}</p>
            ))}
          </DrawerItem>
        )}
        {nodeInfo && (
          <>
            <DrawerItem name="OS">{`${nodeInfo.operatingSystem} (${nodeInfo.architecture})`}</DrawerItem>
            <DrawerItem name="OS Image">{nodeInfo.osImage}</DrawerItem>
            <DrawerItem name="Kernel version">{nodeInfo.kernelVersion}</DrawerItem>
            <DrawerItem name="Container runtime">{nodeInfo.containerRuntimeVersion}</DrawerItem>
            <DrawerItem name="Kubelet version">{nodeInfo.kubeletVersion}</DrawerItem>
          </>
        )}
        {taints.length > 0 && (
          <DrawerItem name="Taints" labelsOnly>
            {taints.map((taint) => (
              <Badge key={taint.key} label={formatNodeTaint(taint)} />
            ))}
          </DrawerItem>
        )}
        {conditions && (
          <DrawerItem name="Conditions" className="conditions" labelsOnly>
            {conditions.map((condition) => (
              <Badge
                key={condition.type}
                label={condition.type}
                className={kebabCase(condition.type)}
                tooltip={{
                  formatters: {
                    tableView: true,
                  },
                  children: Object.entries(condition).map(([key, value]) => (
                    <div key={key} className="flex gaps align-center">
                      <div className="name">{upperFirst(key)}</div>
                      <div className="value">{value}</div>
                    </div>
                  )),
                }}
              />
            ))}
          </DrawerItem>
        )}
        <DrawerTitle>Capacity</DrawerTitle>
        <NodeDetailsResources node={node} type="capacity" />
        <DrawerTitle>Allocatable</DrawerTitle>
        <NodeDetailsResources node={node} type="allocatable" />
        <PodDetailsList
          pods={childPods}
          owner={node}
          maxCpu={node.getCpuCapacity()}
          maxMemory={node.getMemoryCapacity()}
        />
      </div>
    );
  }
}

export const NodeDetails = withInjectables<Dependencies, NodeDetailsProps>(NonInjectedNodeDetails, {
  getProps: (di, props) => ({
    ...props,
    subscribeStores: di.inject(subscribeStoresInjectable),
    podStore: di.inject(podStoreInjectable),
    logger: di.inject(loggerInjectionToken),
    loadPodsFromAllNamespaces: di.inject(loadPodsFromAllNamespacesInjectable),
  }),
});
