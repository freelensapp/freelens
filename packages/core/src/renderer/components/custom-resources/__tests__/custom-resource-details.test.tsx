/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { CustomResourceDefinition, KubeObject } from "@freelensapp/kube-object";
import React from "react";
import { getDiForUnitTesting } from "../../../getDiForUnitTesting";
import type { DiRender } from "../../test-utils/renderFor";
import { renderFor } from "../../test-utils/renderFor";
import { CustomResourceDetails } from "../details";

describe("<CustomResourceDetails />", () => {
  let render: DiRender;

  beforeEach(() => {
    const di = getDiForUnitTesting();

    render = renderFor(di);
  });

  describe("with a CRD with a boolean field", () => {
    let crd: CustomResourceDefinition;

    beforeEach(() => {
      crd = new CustomResourceDefinition({
        apiVersion: "apiextensions.k8s.io/v1",
        kind: "CustomResourceDefinition",
        metadata: {
          name: "my-crd",
          resourceVersion: "1",
          selfLink: "/apis/apiextensions.k8s.io/v1/customresourcedefinitions/my-crd",
          uid: "1",
        },
        spec: {
          versions: [
            {
              name: "v1",
              served: true,
              storage: true,
              schema: {
                openAPIV3Schema: {
                  type: "object",
                  properties: {
                    spec: {
                      type: "object",
                      properties: {
                        "my-field": {
                          type: "boolean",
                        },
                      },
                    },
                  },
                },
              },
              additionalPrinterColumns: [
                {
                  name: "MyField",
                  jsonPath: ".spec.my-field",
                  type: "boolean",
                },
              ],
            },
          ],
          group: "stable.lens.dev",
          names: {
            kind: "MyCrd",
            plural: "my-crds",
          },
          scope: "Cluster",
        },
      });
    });

    it("should display false in an additionalPrinterColumn as 'false'", () => {
      const cr = new KubeObject({
        apiVersion: "stable.lens.dev/v1",
        kind: "MyCrd",
        metadata: {
          name: "first-crd",
          resourceVersion: "1",
          selfLink: "stable.lens.dev/v1/first-crd",
          uid: "2",
        },
        spec: {
          "my-field": false,
        },
      });
      const result = render(<CustomResourceDetails crd={crd} object={cr} />);

      expect(result.container).toMatchSnapshot();
      expect(result.getByText("false")).toBeTruthy();
    });

    it("should display true in an additionalPrinterColumn as 'true'", () => {
      const cr = new KubeObject({
        apiVersion: "stable.lens.dev/v1",
        kind: "MyCrd",
        metadata: {
          name: "first-crd",
          resourceVersion: "1",
          selfLink: "stable.lens.dev/v1/first-crd",
          uid: "2",
        },
        spec: {
          "my-field": true,
        },
      });
      const result = render(<CustomResourceDetails crd={crd} object={cr} />);

      expect(result.container).toMatchSnapshot();
      expect(result.getByText("true")).toBeTruthy();
    });
  });

  describe("with a CRD with a number field", () => {
    let crd: CustomResourceDefinition;

    beforeEach(() => {
      crd = new CustomResourceDefinition({
        apiVersion: "apiextensions.k8s.io/v1",
        kind: "CustomResourceDefinition",
        metadata: {
          name: "my-crd",
          resourceVersion: "1",
          selfLink: "/apis/apiextensions.k8s.io/v1/customresourcedefinitions/my-crd",
          uid: "1",
        },
        spec: {
          versions: [
            {
              name: "v1",
              served: true,
              storage: true,
              schema: {
                openAPIV3Schema: {
                  type: "object",
                  properties: {
                    spec: {
                      type: "object",
                      properties: {
                        "my-field": {
                          type: "number",
                        },
                      },
                    },
                  },
                },
              },
              additionalPrinterColumns: [
                {
                  name: "MyField",
                  jsonPath: ".spec.my-field",
                  type: "number",
                },
              ],
            },
          ],
          group: "stable.lens.dev",
          names: {
            kind: "MyCrd",
            plural: "my-crds",
          },
          scope: "Cluster",
        },
      });
    });

    it("should display 0 in an additionalPrinterColumn as '0'", () => {
      const cr = new KubeObject({
        apiVersion: "stable.lens.dev/v1",
        kind: "MyCrd",
        metadata: {
          name: "first-crd",
          resourceVersion: "1",
          selfLink: "stable.lens.dev/v1/first-crd",
          uid: "2",
        },
        spec: {
          "my-field": 0,
        },
      });
      const result = render(<CustomResourceDetails crd={crd} object={cr} />);

      expect(result.container).toMatchSnapshot();
      expect(result.getByText("0")).toBeTruthy();
    });

    it("should display 1234 in an additionalPrinterColumn as '1234'", () => {
      const cr = new KubeObject({
        apiVersion: "stable.lens.dev/v1",
        kind: "MyCrd",
        metadata: {
          name: "first-crd",
          resourceVersion: "1",
          selfLink: "stable.lens.dev/v1/first-crd",
          uid: "2",
        },
        spec: {
          "my-field": 1234,
        },
      });
      const result = render(<CustomResourceDetails crd={crd} object={cr} />);

      expect(result.container).toMatchSnapshot();
      expect(result.getByText("1234")).toBeTruthy();
    });
  });
});
