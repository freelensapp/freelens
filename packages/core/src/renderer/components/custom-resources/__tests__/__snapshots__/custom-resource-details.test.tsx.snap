// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomResourceDetails /> with a CRD with a boolean field should display false in an additionalPrinterColumn as 'false' 1`] = `
<div>
  <div
    class="CustomResourceDetails MyCrd"
  >
    <div
      class="DrawerItem"
    >
      <span
        class="name"
      >
        My Field
      </span>
      <span
        class="value"
      >
        <div
          class="badge false"
        >
          False
        </div>
      </span>
    </div>
  </div>
</div>
`;

exports[`<CustomResourceDetails /> with a CRD with a boolean field should display true in an additionalPrinterColumn as 'true' 1`] = `
<div>
  <div
    class="CustomResourceDetails MyCrd"
  >
    <div
      class="DrawerItem"
    >
      <span
        class="name"
      >
        My Field
      </span>
      <span
        class="value"
      >
        <div
          class="badge true"
        >
          True
        </div>
      </span>
    </div>
  </div>
</div>
`;

exports[`<CustomResourceDetails /> with a CRD with a number field should display 0 in an additionalPrinterColumn as '0' 1`] = `
<div>
  <div
    class="CustomResourceDetails MyCrd"
  >
    <div
      class="DrawerItem"
    >
      <span
        class="name"
      >
        My Field
      </span>
      <span
        class="value"
      >
        0
      </span>
    </div>
  </div>
</div>
`;

exports[`<CustomResourceDetails /> with a CRD with a number field should display 1234 in an additionalPrinterColumn as '1234' 1`] = `
<div>
  <div
    class="CustomResourceDetails MyCrd"
  >
    <div
      class="DrawerItem"
    >
      <span
        class="name"
      >
        My Field
      </span>
      <span
        class="value"
      >
        1234
      </span>
    </div>
  </div>
</div>
`;
